Add CanvasRenderingContext2DSettings.pixelFormat to support higher bit depths

squelart opened the pull request:
#### e9af57b86e4d575626553d667fa60595ffbd688c
<pre>
Add CanvasRenderingContext2DSettings.pixelFormat to support higher bit depths
<a href="https://bugs.webkit.org/show_bug.cgi?id=283281">https://bugs.webkit.org/show_bug.cgi?id=283281</a>
<a href="https://rdar.apple.com/problem/140105324">rdar://problem/140105324</a>

Reviewed by Simon Fraser.

Start supporting different pixel formats in 2D canvas, as per
<a href="https://github.com/w3c/ColorWeb-CG/blob/main/canvas_float.md">https://github.com/w3c/ColorWeb-CG/blob/main/canvas_float.md</a> ,
with naming based on Chromium implementation.
This is needed to eventually support HDR.

This adds:
- An unstable preference &quot;CanvasPixelFormatEnabled&quot;.
- CanvasRenderingContext2DSettings.pixelFormat of type CanvasPixelFormat,
  which can be &quot;uint8&quot; (default) or &quot;float16&quot; for now.
  (The property has no effects yet, but can be set and queried.)

* LayoutTests/fast/canvas/CanvasRenderingContext2DSettings-pixelFormat-disabled-expected.txt: Added.
* LayoutTests/fast/canvas/CanvasRenderingContext2DSettings-pixelFormat-disabled.html: Added.
* LayoutTests/fast/canvas/CanvasRenderingContext2DSettings-pixelFormat-enabled-expected.txt: Added.
* LayoutTests/fast/canvas/CanvasRenderingContext2DSettings-pixelFormat-enabled.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/canvas/CanvasRenderingContext2DSettings.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DSettings.idl:

Canonical link: <a href="https://commits.webkit.org/286934@main">https://commits.webkit.org/286934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bda44123a67d132fdd265b6056ea1a2d04920a85

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77506 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56541 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82090 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28788 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79623 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65690 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4838 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60729 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18728 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80573 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66527 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41007 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48105 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24023 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27116 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70694 "Found 5 new JSC stress test failures: stress/get-by-val-hoist-above-structure-2.js.ftl-eager-no-cjit-b3o1, stress/get-by-val-hoist-above-structure-2.js.ftl-no-cjit-small-pool, stress/get-by-val-hoist-above-structure.js.ftl-eager-no-cjit-b3o1, stress/get-by-val-hoist-above-structure.js.ftl-no-cjit-no-put-stack-validate, wasm.yaml/wasm/modules/wasm-imports-js-exports.js.wasm-no-cjit (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69215 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24361 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83495 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76785 "Found 3 new JSC stress test failures: stress/attribute-custom-accessor.js.ftl-eager, stress/get-by-val-hoist-above-structure.js.ftl-eager-no-cjit-b3o1, wasm.yaml/wasm/stress/referenced-function.js.wasm-collect-continuously (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4886 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3327 "Found 3 new test failures: imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-on-dialog-invalid-behavior.tentative.html ipc/create-connection-and-send-async.html webrtc/vp9-profile2.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68966 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5042 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66494 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68231 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17063 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12248 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10346 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99037 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4833 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7648 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21650 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4852 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8287 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6611 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/html/canvas/CanvasDataType.h:
```diff
@@ -0,0 +1,32 @@
+
+namespace WebCore {
+
+enum class CanvasDataType { Unorm8, Float16 };
```
` : bool ` ?

squelart replied:
I personally don't like `bool` for things that don't really mean false/true.
Also, in this case it could well be extended in the near future (e.g., Float32 is an option in the proposed specs).
So how about `uint8_t`? It's as compact, and IPCs are still checked to only accept the possible enum values.

weinig replied:
`: bool` is what we currently use for these kind single bit things. If you think WebCore overall should stop using `bool`, that should be raised globally. Having one part of the code base diverge due to developer preference doesn't make a lot of sense for a project of this size.

squelart replied:
There are hundreds of instances of uint8_t-based enums with two enumerations already, so it's not uncommon.
Anyway, two votes against, I'll change to bool...

weinig commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -1776,6 +1776,22 @@ CanvasFingerprintingQuirkEnabled:
     WebCore:
       default: true
 
+CanvasHDREnabled:
```
I would really love to avoid using the term HDR in WebCore as it quite vague, and have steered clear of it in the color work so far. 

For this setting, perhaps CanvasFloat16BackingStore or CanvasBackingStoreDataTypes would work?

squelart replied:
This patch here is only part 1 of N to implement https://github.com/w3c/ColorWeb-CG/blob/main/hdr_html_canvas_element.md , I was going to use this setting for the added colorspaces as well.

We've used "HDR" for other recent developments (HDRForImagesEnabled and WebGPUHDREnabled); it doesn't make it right, but that boat may have already sailed!

Would you have other suggestions for a setting related to "Adding support High Dynamic Range (HDR) imagery to HTML Canvas"?

squelart replied:
Should I move this new setting to a separate patch, to avoid confusing this feature-wide change from the small-part CanvasDataType addition?

weinig replied:
I think having two settings, one for the pixel formats and one for the color spaces is fine. 

Is this stuff really ready to be implemented? It doesn’t seem like a very complete spec, but perhaps I am misreading. 

Seems a little weird for canvas to have such a disjoint set of color spaces from CSS for instance.

weinig replied:
> Would you have other suggestions for a setting related to "Adding support High Dynamic Range (HDR) imagery to HTML Canvas"?

yeah, I would prefer those say what they were actually doing, which was probably support for specific color spaces/transfer encoding?

The fact that I can’t tell what it means is exactly the reason I am pushing to not use the term in WebKit.

weinig replied:
> Is this stuff really ready to be implemented? It doesn’t seem like a very complete spec, but perhaps I am misreading.
> 
> Seems a little weird for canvas to have such a disjoint set of color spaces from CSS for instance.

Reading this back, it sounds like I am pushing you not to work on this, which is not my intention. I am curious about why the spec is not in the usually W3C spec form, but this is sometimes how things go.

squelart replied:
Alright, I've renamed the setting `CanvasPixelFormatEnabled`, I can see the benefit now. Thank you for your review.

The feature is indeed lacking a proper spec, so it may still be unstable (as the setting denotes, and it's off by default). Hopefully ColorWeb-CG can work on it soon...
Also I've changed the names to match the current Chrome implementation (mainly dataType -> pixelFormat), to be able to compare the results from some early-adopter websites; this naming may still change of course.

squelart replied:
Feel free to comment at https://github.com/whatwg/html/issues/9461 if you're happy to share your opinions and suggestions there 😉

weinig commented on Source/WebCore/html/canvas/CanvasDataType.idl:
```diff
@@ -0,0 +1,26 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+enum CanvasDataType { "unorm8", "float16" };
```
Please add a link to the spec with this IDL addition above this declaration. This is the custom in IDL file to make it easier to find the right original source to compare with.

squelart replied:
Thanks for the reminder!

squelart committed:
Add CanvasRenderingContext2DSettings.pixelFormat to support higher bit depths
https://bugs.webkit.org/show_bug.cgi?id=283281
rdar://problem/140105324

Reviewed by Simon Fraser.

Start supporting different pixel formats in 2D canvas, as per
https://github.com/w3c/ColorWeb-CG/blob/main/canvas_float.md ,
with naming based on Chromium implementation.
This is needed to eventually support HDR.

This adds:
- An unstable preference "CanvasPixelFormatEnabled".
- CanvasRenderingContext2DSettings.pixelFormat of type CanvasPixelFormat,
  which can be "uint8" (default) or "float16" for now.
  (The property has no effects yet, but can be set and queried.)

* LayoutTests/fast/canvas/CanvasRenderingContext2DSettings-pixelFormat-disabled-expected.txt: Added.
* LayoutTests/fast/canvas/CanvasRenderingContext2DSettings-pixelFormat-disabled.html: Added.
* LayoutTests/fast/canvas/CanvasRenderingContext2DSettings-pixelFormat-enabled-expected.txt: Added.
* LayoutTests/fast/canvas/CanvasRenderingContext2DSettings-pixelFormat-enabled.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/canvas/CanvasRenderingContext2DSettings.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DSettings.idl:

Canonical link: https://commits.webkit.org/286934@main
